### PR TITLE
DDF-2508 added ability to specify individual override attributes in REST request

### PR DESCRIPTION
**Pull request by stustison**
#### Who is reviewing it (please choose AT LEAST two reviewers that need to approve the PR before it can get merged; if a component team is listed, at least one of its members needs to approve)?
@garrettfreibott @coyotesqrl @rzwiefel 
#### Choose 2 committers to review/merge the PR (please choose ONLY two committers from below, delete the rest).
@pklinef
@stustison
#### How should this be tested?
Full build + manual test of resource and parsed metadata
#### Checklist:
- [x] Documentation Updated
- [x] Update / Add Unit Tests
- [ ] Update / Add Integration Tests

 - added option to parse metadata
 - fixed bug with derived resources possibly overriding the resource uri of non-derived resources

### CODE REVIEW

File: catalog/rest/catalog-rest-endpoint/src/main/java/org/codice/ddf/endpoints/rest/RESTEndpoint.java [hunk 1]
```diff
@@ -836,9 +858,128 @@ public Response addDocument(@Context HttpHeaders headers, @Context UriInfo reque
...
+            Metacard metacard = null;
+            CreateInfo createInfo = null;
+
+            for (Attachment attachment : contentParts) {
+                String name = attachment.getContentDisposition()
+                        .getParameter("name");
+
+                String parsedName = (name.startsWith("parse.")) ? name.substring(6) : name;
+                try {
+                    InputStream inputStream = attachment.getDataHandler()
```
Comment (garrettfreibott) [inline catalog/rest/catalog-rest-endpoint/src/main/java/org/codice/ddf/endpoints/rest/RESTEndpoint.java]: Would a try-with-resources work here? This would eliminate the need to manually close the input stream in all the `if` cases.

File: catalog/rest/catalog-rest-endpoint/src/main/java/org/codice/ddf/endpoints/rest/RESTEndpoint.java [hunk 2]
```diff
@@ -836,9 +858,128 @@ public Response addDocument(@Context HttpHeaders headers, @Context UriInfo reque
...
+                                    "assigned-when-ingested", inputStream, transformer);
+                        } catch (MimeTypeParseException | MetacardCreationException e) {
+                            LOGGER.debug("Unable to parse metadata {}",
+                                    attachment.getContentType()
+                                            .toString());
+                        } finally {
+                            IOUtils.closeQuietly(inputStream);
+                        }
+
+                    } else if (name.equals("parse.resource")) {
```
Comment (stustison) [inline catalog/rest/catalog-rest-endpoint/src/main/java/org/codice/ddf/endpoints/rest/RESTEndpoint.java]: No, it won't be closed in this method and that is the reason why I couldn't use a try with resources. It needs to be available in separate threads as the request passes through the framework. The other stuff I can read and drop into attributes, but the content is potentially huge.

File: catalog/rest/catalog-rest-endpoint/src/main/java/org/codice/ddf/endpoints/rest/RESTEndpoint.java (hunk 1 above)
Comment (stustison) [inline catalog/rest/catalog-rest-endpoint/src/main/java/org/codice/ddf/endpoints/rest/RESTEndpoint.java]: Nope, see below.

File: catalog/rest/catalog-rest-endpoint/src/main/java/org/codice/ddf/endpoints/rest/RESTEndpoint.java [hunk 3]
```diff
@@ -836,9 +858,128 @@ public Response addDocument(@Context HttpHeaders headers, @Context UriInfo reque
         return response;
     }
 
+    CreateInfo parseAttachments(List<Attachment> contentParts, String transformerParam) {
+
+        if (contentParts.size() == 1) {
+            Attachment contentPart = contentParts.get(0);
+            return parseAttachment(contentPart);
+        } else {
```
Comment (stustison) [inline catalog/rest/catalog-rest-endpoint/src/main/java/org/codice/ddf/endpoints/rest/RESTEndpoint.java]: Sure.

File: catalog/rest/catalog-rest-endpoint/src/main/java/org/codice/ddf/endpoints/rest/RESTEndpoint.java [hunk 4]
```diff
@@ -836,9 +858,128 @@ public Response addDocument(@Context HttpHeaders headers, @Context UriInfo reque
...
+            List<Attribute> attributes = new ArrayList<>(contentParts.size());
+            Metacard metacard = null;
+            CreateInfo createInfo = null;
+
+            for (Attachment attachment : contentParts) {
+                String name = attachment.getContentDisposition()
+                        .getParameter("name");
+
+                String parsedName = (name.startsWith("parse.")) ? name.substring(6) : name;
```
Comment (pklinef) [inline catalog/rest/catalog-rest-endpoint/src/main/java/org/codice/ddf/endpoints/rest/RESTEndpoint.java]: Could use `StringUtils.substringAfter`.

https://commons.apache.org/proper/commons-lang/apidocs/org/apache/commons/lang3/StringUtils.html#substringAfter(java.lang.String,%20java.lang.String)

File: catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/impl/operations/UpdateOperations.java [hunk 5]
```diff
@@ -647,4 +670,45 @@ private void logFailedQueryInfo(UpdateRequest updateRequest, QueryResponse query
...
+        for (ContentItem contentItem : updateStorageRequest.getContentItems()) {
+            Metacard metacard = metacardMap.get(contentItem.getId());
+            MetacardType metacardType = metacard.getMetacardType();
+
+            if (contentItem.getMetacard() != null) {
+                Metacard overrideMetacard = contentItem.getMetacard();
+                MetacardType overrideMetacardType = contentItem.getMetacard()
+                        .getMetacardType();
+
+                if (overrideMetacardType.equals(BasicTypes.BASIC_METACARD)) {
```
Comment (pklinef) [inline catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/impl/operations/UpdateOperations.java]: You knew this comment was coming but I think you can extract this into a class that extends MetacardImpl or create a factory that merges overrides.  That class can just live in the standard framework project.  Then both operations can use this common logic.

File: distribution/docs/src/main/resources/_contents/_catalog-contents/integrating-catalog-contents.adoc [hunk 6]
```diff
@@ -179,6 +182,10 @@ Note that for all RESTful CRUD commands only one metacard ID is supported in t
 Any web browser can be used to perform a REST read.
 Various other tools and libraries can be used to perform the other HTTP operations on the REST endpoint (e.g., soapUI, cURL, etc.)
 
+The REST endpoint can be used to upload resources as attachments. The `create` and `update` methods both support the multipart mime format. If only a single attachment exists, it will be interpreted as a resource to be parsed, which will result in a metacard and resource being stored in the system.
+
+If multiple attachments exist, then the REST endpoint will assume that 1 attachment is the actual resource (attachment should be named `parse.resource`) and the other attachments are overrides of metacard attributes (attachment names should follow metacard attribute names). In the case of the metadata attribute, it is possible to also have the system transform that metadata and use the results of that to override the metacard that would be generated from the resource (attachment should be named `parse.metadata`).
```
Comment (pklinef) [inline distribution/docs/src/main/resources/_contents/_catalog-contents/integrating-catalog-contents.adoc]: It would be helpful to provide a example request of how to do this.

File: catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/impl/operations/CreateOperations.java [hunk 7]
```diff
@@ -683,6 +697,43 @@ private CreateStorageRequest applyAttributeOverrides(CreateStorageRequest create
...
+                    overrideMetacardType.getAttributeDescriptors()
+                            .stream()
+                            .map(attributeDescriptor -> overrideMetacard.getAttribute(
+                                    attributeDescriptor.getName()))
+                            .filter(Objects::nonNull)
+                            .filter(attribute -> !attribute.getName()
+                                    .equals(Metacard.ID))
+                            .filter(attribute -> !attribute.getName()
+                                    .equals(Metacard.RESOURCE_URI))
+                            .forEach(metacard::setAttribute);
```
Comment (pklinef) [inline catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/impl/operations/CreateOperations.java]: I think you can rewrite this without side effects so that you do not need this if-else block and largely duplicate code.

**Comment by rzwiefel** [comment]
TL;DR:
metacardtype descriptors -> get off metacard -> filter nulls -> filter out ID or resource URI -> override. 

I'm only cheating because all i can see are phils comment staring back at me, but definitely at least refactoring that logic to a common place would be :ok_hand:

File: distribution/docs/src/main/resources/_contents/_catalog-contents/integrating-catalog-contents.adoc (hunk 6 above)
Comment (stustison) [inline distribution/docs/src/main/resources/_contents/_catalog-contents/integrating-catalog-contents.adoc]: OK.

File: catalog/rest/catalog-rest-endpoint/src/main/java/org/codice/ddf/endpoints/rest/RESTEndpoint.java [hunk 8]
```diff
@@ -777,14 +799,14 @@ public Response addDocument(@Context HttpHeaders headers, @Context UriInfo reque
                 if (createInfo == null) {
                     CreateRequest createRequest = new CreateRequestImpl(generateMetacard(mimeType,
                             null,
-                            message));
+                            message, transformerParam));
```
Comment (coyotesqrl) [inline catalog/rest/catalog-rest-endpoint/src/main/java/org/codice/ddf/endpoints/rest/RESTEndpoint.java]: I think this needs to be reformatted.

File: catalog/rest/catalog-rest-endpoint/src/main/java/org/codice/ddf/endpoints/rest/RESTEndpoint.java [hunk 9]
```diff
@@ -836,9 +858,126 @@ public Response addDocument(@Context HttpHeaders headers, @Context UriInfo reque
...
+                        LOGGER.debug("Unable to parse metadata {}",
+                                attachment.getContentType()
+                                        .toString());
+                    } finally {
+                        IOUtils.closeQuietly(inputStream);
+                    }
+
+                } else if (name.equals("parse.resource")) {
+                    createInfo = parseAttachment(attachment);
+                } else {
```
Comment (coyotesqrl) [inline catalog/rest/catalog-rest-endpoint/src/main/java/org/codice/ddf/endpoints/rest/RESTEndpoint.java]: I would probably read more cleanly if you extracted methods for the if and else blocks here.

File: catalog/rest/catalog-rest-endpoint/src/main/java/org/codice/ddf/endpoints/rest/RESTEndpoint.java [hunk 10]
```diff
@@ -836,9 +858,126 @@ public Response addDocument(@Context HttpHeaders headers, @Context UriInfo reque
...
+                                "assigned-when-ingested", inputStream, transformer);
+                    } catch (MimeTypeParseException | MetacardCreationException e) {
+                        LOGGER.debug("Unable to parse metadata {}",
+                                attachment.getContentType()
+                                        .toString());
+                    } finally {
+                        IOUtils.closeQuietly(inputStream);
+                    }
+
+                } else if (name.equals("parse.resource")) {
```
Comment (coyotesqrl) [inline catalog/rest/catalog-rest-endpoint/src/main/java/org/codice/ddf/endpoints/rest/RESTEndpoint.java]: If you switched the order of these first two tests, you could defer the retrieval of the inputStream a bit. Not a huge improvement, but it would be a bit cleaner and more clear.

File: catalog/rest/catalog-rest-endpoint/src/main/java/org/codice/ddf/endpoints/rest/RESTEndpoint.java [hunk 11]
```diff
@@ -836,9 +858,126 @@ public Response addDocument(@Context HttpHeaders headers, @Context UriInfo reque
...
+                    createInfo = parseAttachment(attachment);
+                } else {
+                    Optional<AttributeDescriptor> attributeDescriptor = metacardTypes.stream()
+                            .map(metacardType -> metacardType.getAttributeDescriptor(
+                                    parsedName))
+                            .filter(Objects::nonNull)
+                            .findFirst();
```
Comment (coyotesqrl) [inline catalog/rest/catalog-rest-endpoint/src/main/java/org/codice/ddf/endpoints/rest/RESTEndpoint.java]: Adding a `.map(AttributeDescriptor::getType).map(AttributeType::getAttributeFormat)` here would simplify your switch.

File: distribution/docs/src/main/resources/_contents/_catalog-contents/integrating-catalog-contents.adoc [hunk 12]
```diff
@@ -100,17 +100,20 @@ The RESTful CRUD Endpoint provides the capability to query, create, update, and
 |`create`
 |HTTP POST
 |HTTP request body contains the input to be ingested.
-|\http://<DISTRIBUTION_HOST>:<DISTRIBUTION_PORT>/services/catalog
+|\http://<DISTRIBUTION_HOST>:<DISTRIBUTION_PORT>/services/catalog?transform=<input transformer>
```
Comment (coyotesqrl) [inline distribution/docs/src/main/resources/_contents/_catalog-contents/integrating-catalog-contents.adoc]: Should indicated that this new query param is optional.

File: catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/impl/operations/OverrideAttributesSupport.java [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+import ddf.catalog.data.Metacard;
+import ddf.catalog.data.MetacardType;
+import ddf.catalog.data.impl.BasicTypes;
+
+public class OverrideAttributesSupport {
+    private OverrideAttributesSupport() {
+
+    }
+
+    public static void overrideAttributes(List<ContentItem> contentItems,
```
Comment (pklinef) [inline catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/impl/operations/OverrideAttributesSupport.java]: I think you can use the wrapped metacard capability of `MetacardImpl` to make this a little more readable.

```Java
    public static void overrideAttributes(List<ContentItem> contentItems,
            Map<String, Metacard> metacardMap) {
        for (ContentItem contentItem : contentItems) {
            if (contentItem.getMetacard() == null) {
                continue;
            }

            Metacard currentMetacard = metacardMap.get(contentItem.getId());
            Metacard overrideMetacard = contentItem.getMetacard();

            MetacardType updatedMetacardType;
            if (!BasicTypes.BASIC_METACARD.equals(overrideMetacard.getMetacardType())) {
                updatedMetacardType = overrideMetacard.getMetacardType();
            } else {
                updatedMetacardType = currentMetacard.getMetacardType();
            }

            Metacard updatedMetacard = new MetacardImpl(currentMetacard, updatedMetacardType);

            addAttributes(updatedMetacard, overrideMetacard);

            metacardMap.put(contentItem.getId(), updatedMetacard);
        }
    }

    private static void addAttributes(Metacard metacard, Metacard otherMetacard) {
        otherMetacard.getMetacardType().getAttributeDescriptors()
                .stream()
                .map(attributeDescriptor -> otherMetacard.getAttribute(
                        attributeDescriptor.getName()))
                .filter(Objects::nonNull)
                .filter(attribute -> !attribute.getName()
                        .equals(Metacard.ID))
                .filter(attribute -> !attribute.getName()
                        .equals(Metacard.RESOURCE_URI))
                .forEach(metacard::setAttribute);
    }
```

Comment (stustison) [inline catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/impl/operations/OverrideAttributesSupport.java]: sgtm, I forgot about the wrapped metacard stuff

File: catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/impl/operations/OverrideAttributesSupport.java [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+            Map<String, Metacard> metacardMap) {
+        for (ContentItem contentItem : contentItems) {
+            if (contentItem.getMetacard() == null) {
+                continue;
+            }
+
+            Metacard currentMetacard = metacardMap.get(contentItem.getId());
```
Comment (pklinef) [inline catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/impl/operations/OverrideAttributesSupport.java]: If you extract the code from here to line 50 into a public method that takes two metacards, I think you might be able to remove the duplicate code from `CreateRequest.populateMetacardMap` and `UpdateResponseupdate`.

File: distribution/docs/src/main/resources/_contents/_catalog-contents/integrating-catalog-contents.adoc [hunk 15]
```diff
@@ -179,6 +182,31 @@ Note that for all RESTful CRUD commands only one metacard ID is supported in t
...
+
+
+------WebKitFormBoundary7MA4YWxkTrZu0gW
+Content-Disposition: form-data; name="parse.metadata"; filename=""
+Content-Type:
+
+
+------WebKitFormBoundary7MA4YWxkTrZu0gW--
+----
+
```
Comment (ricklarsen) [inline distribution/docs/src/main/resources/_contents/_catalog-contents/integrating-catalog-contents.adoc]: I'm copying these changes into the draft-documentation version too, so they are not lost when we transition.
